proxy/proxy/client.go

```diff
@@ -292,33 +240,126 @@ func isExpired(cfg *tls.Config) bool {
 	return time.Now().After(cfg.Certificates[0].Leaf.NotAfter)
 }
 
-func (c *Client) cachedCfg(instance string) (string, *tls.Config, string) {
+// startRefresh kicks off a refreshCfg asynchronously, returning a channel that is closed
+// when it's complete and updates the cacheEntry when complete. This function should only
+// be called from the scope of "cachedCfg", which controls the logic around throttling refreshes.
+func (c *Client) startRefresh(instance string, refreshCfgBuffer time.Duration) chan bool {
+	done := make(chan bool)
+	go func() {
+		defer close(done)
+		addr, cfg, ver, err := c.refreshCfg(instance)
+
+		c.cacheL.Lock()
+		old := c.cfgCache[instance]
+		// if we failed to refresh cfg do not throw out potentially valid one
+		if err != nil && !isExpired(old.cfg) {
+			logging.Errorf("failed to refresh the ephemeral certificate for %s, returning previous cert instead: %v", instance, err)
+			addr, cfg, ver, err = old.addr, old.cfg, old.version, old.err
+		}
+		c.cfgCache[instance] = cacheEntry{
+			lastRefreshed: time.Now(),
+			err:           err,
+			addr:          addr,
+			version:       ver,
+			cfg:           cfg,
+			done:          done,
+		}
+		c.cacheL.Unlock()
+
+		certExpiration := cfg.Certificates[0].Leaf.NotAfter
```

Comment:
I'm not sure I understand what your suggestion is. Are you suggesting moving it into `refreshCfg` or just scoping to `startRefresh`? 

You don't want to move it into `refreshCfg`, because you want to update the cacheEntry and schedule the refresh again, regardless of how the last operation went.

You don't want it to move scope up into `startRefresh` because you don't have the results of the refresh operation yet.

Reply:
I understand this code better now. Let me pose a different question below. Resolving.